### ISSUE TITLE
Question: Using microsoft configure<T> option methods with child containers causes root container configure option delegates to be invokes after child.

### ISSUE THREAD

**Issue by vladimir-ilnytskyi** [z4kn4fein/stashbox#182]
I use custom multitenancy where i have some common confituration which is manged in root container and for each tenant individually I have tenant specific configs and/or overrides of parent.

Using microsoft configure<T> option methods with child containers causes root configure option delegates to be invokes after child delegates in child container.

// registrations
Having stashbox configured with root (container using microsoft interfaces via service provider factory but i assume that same would apply for stashbox plain interfaces)

// register parent options
rootservices.Configure<Foo>(<options name>, fooConfigSection);

// register child (happens in multitenancy middleware on tenant configuration initialization)
childContainerServices.Configure<Foo>(<options name>, childFooConfigSection);

// resolutions
var optionsInstance = childContainer.GetRequiredService<IOptionsMonitor<Foo>>().Get(<options name>); // here i get effectively parent options instead of child options.

I assume this is because OptionsFactory<T> has c-tor accepting IEnumerable<IConfigureOptions<T>> which makes child delegates to have higher priority then parent so they are resolved prior to parent in ienumerable.

There are workarounds obviously like not using same type and/or options name but out of curiosity what are your suggestions how to handle this behavior?

Thanks

**Comment by z4kn4fein** [comment]
Hi @vladimir-ilnytskyi, thank you for reaching out! I'm not sure I fully understand the issue, could you please share a whole snippet that describes the problem? Thanks!

**Comment by vladimir-ilnytskyi** [comment]
Sure! Here i'll provide code for multitenant container - basically my implementation of dynamic multitenancy ( where tenant configurations are stored on remote source and can be added dynamically and also they change dynamically, so i will also skip here some pieces like how i get IConfiguration from tenant object... 


- Approx code snippet in program.cs (.net8) : 

```
var services = builder.Services;
var configuration = builder.Configuration;

// public class Foo { public string Property1 {get; set;} public string Property2 {get; set;} }

services.Configure<Foo>(configuration.GetSection("Foo")); // configure in parent container - e.g. root configuration is from appsettings and it binds Property1 here, which i want to override for some tenants from their configuration a bit below 

var tenantContainerConfigureAction = (IServiceCollection tenantServices, IConfiguration tenantConfiguration) => {

// tenantConfiguration - is completely isolated remote config that has some stuff that i want to override/add so instances of IOptionsMonitor<Foo> give me updated value for that tenant

tenantServices.Configure<Foo>(tenantConfiguration.GetSection("Foo")); // configure in tenant (child) container. Here i want to override Property1 from Foo section

....register some services etc...
};

builder.UseServiceProviderFactory(context =>
        {
            var rootContainer = new StashboxContainer();
            if (context.HostingEnvironment.IsDevelopment())
            {
                rootContainer.Configure(config => config.WithLifetimeValidation());
            }

            return new MultiTenantServiceProviderFactory(new MultiTenantContainer(tenantContainerConfigureAction, rootContainer));
        });

```


- Problem 

when i'm already in controller and using child container services when i resolve 
IOptionsMonitor<Foo> optionsMonitor and get my instance of options

optionsMonitor.CurrentValue - this one is constructed by OptionsFactory<Foo> (microsoft options class) with not expected for me order. This OptionsFactory<T> has c-tor that accepts IEnumerable<IConfigureOptions<T>>

So it first runs child configure delegates, then parent...
I guess stashbox resolves all registrations of IConfigureOptions<T> if it is  IEnumerable<IConfigureOptions<T>> combining them - first from child then from parent.

<img width="936" alt="image" src="https://github.com/user-attachments/assets/54c79136-4eee-4cfc-a99f-1f4f856f7a5d" />
 


- Some classes for general understanding... 


Middleware that swaps request services in httpcontext based on current tenant
```
public async Task InvokeAsync(HttpContext context, RequestDelegate next)
    {
        // if current request context is configured to ignore multitenancy call the next middleware
        if (_options.Value.IgnoreMultitenancy?.Invoke(context) ?? false)
        {
            await next(context);
            return;
        }

        var tenantResolutionStrategy = _tenantResolutionStrategy;
        var endpointMeta = context.GetEndpoint()?.Metadata;
        if (endpointMeta != null)
        {
            var isMultitenancyDisabled = endpointMeta?.Any(m => m is DisableMultitenancyAttribute);
            if (isMultitenancyDisabled ?? false)
            {
                await next(context);
                return;
            }

            var overrideResolutionStrategyType = endpointMeta?.Select(meta =>
            {
                if (meta is TenantResolutionStrategyOverrideAttibute tenantResolutionStrategyAttribute)
                {
                    return tenantResolutionStrategyAttribute.ResolutionStrategyType;
                }

                return null;
            }).FirstOrDefault(x => x != null);

            if (overrideResolutionStrategyType != null)
            {
                tenantResolutionStrategy = (ITenantResolutionStrategy)context.RequestServices.GetService(overrideResolutionStrategyType) ?? _tenantResolutionStrategy;
            }
        }

        var tenantId = await tenantResolutionStrategy.GetTenantIdAsync(context.RequestAborted);
        if (string.IsNullOrWhiteSpace(tenantId))
        {
            var error = new { Error = "Tenant ID is not provided." };
            _logger.LogInformation(error.Error);
            context.Response.StatusCode = StatusCodes.Status400BadRequest;
            await context.Response.WriteAsJsonAsync(error);
            return;
        }

        if (_tenantAccessor.Tenant == null)
        {
            var tenant = await _tenantStore.GetOrRetrieveAsync(tenantId, context.RequestAborted);
            if (tenant != null)
            {
                _tenantAccessor.Tenant = tenant;
            }
            else
            {
                context.Response.StatusCode = StatusCodes.Status404NotFound;
                var noTenantError = new { Error = $"No tenant was found with ID:{tenantId}" };
                _logger.LogInformation(noTenantError.Error);
                await context.Response.WriteAsJsonAsync(noTenantError);
                return;
            }

            if (!tenant.TryGetConfigurationForService(out _))
            {
                var error = $"Tenant {tenant.Label} with ID: {tenant.Id} has no configuration for service: {Tenant.AssemblyName}";
                throw new TenantServiceConfigurationMissingException(error);
            }
        }

        var tenantContainer = _distributor.GetTenantContainer(_tenantAccessor);
        if (tenantContainer == null)
        {
            await next(context);
            return;
        }

        IServiceProvidersFeature? originalFeature = null;
        try
        {
            var feature = new RequestServicesFeature(context, new StashboxServiceScopeFactory(tenantContainer));

            context.Response.RegisterForDisposeAsync(feature);

            originalFeature = context.Features.Get<IServiceProvidersFeature>();
            context.Features.Set<IServiceProvidersFeature>(feature);

            await next(context);
        }
        finally
        {
            context.Features.Set(originalFeature);
        }
    }
```
    

```
public class MultiTenantContainer : IMultiTenantContainer
{
    private int _disposed;
    private readonly object _sync = new();

    //This is the base application container
    public IStashboxContainer RootContainer { get; }

    //This action configures a container builder
    private readonly Action<IServiceCollection, IConfiguration> _configureTenantTenantContainer;

    public MultiTenantContainer(Action<IServiceCollection, IConfiguration> configureTenantContainer, IStashboxContainer rootContainer = null)
    {
        RootContainer = rootContainer ?? new StashboxContainer();
        _configureTenantTenantContainer = configureTenantContainer;
    }

    public IDependencyResolver? GetTenantContainer(ITenantAccessor tenantAccessor)
    {
        var tenant = tenantAccessor?.Tenant;
        if (tenant == null)
        {
            return null;
        }

        var tenantId = tenant.Id.ToString();

        var tenantContainer = RootContainer.GetChildContainer(tenantId);
        if (tenantContainer != null)
        {
            return tenantContainer;
        }

        lock (_sync)
        {
            tenantContainer = RootContainer.GetChildContainer(tenantId);
            if (tenantContainer != null)
            {
                return tenantContainer;
            }

            var services = new ServiceCollection();
            var tenantConfiguration = tenant.GetServiceTenantConfiguration();
            _configureTenantTenantContainer(services, tenantConfiguration);

            tenantContainer = RootContainer.CreateChildContainer(tenantId, attachToParent: true);
            services.AddSingleton(tenantContainer);
            services.CreateBuilder(tenantContainer);
            tenantContainer.Validate();
        }

        return tenantContainer;
    }

    /// <inheritdoc />
    public void Dispose()
    {
        if (Interlocked.CompareExchange(ref _disposed, 1, 0) != 0)
        {
            return;
        }

        RootContainer.Dispose();
    }

    /// <inheritdoc />
    public ValueTask DisposeAsync()
    {
        if (Interlocked.CompareExchange(ref _disposed, 1, 0) != 0)
        {
            return new ValueTask(Task.CompletedTask);
        }

        return RootContainer.DisposeAsync();
    }
}

```


```
public class MultiTenantServiceProviderFactory : IServiceProviderFactory<IMultiTenantContainer>
{
    private readonly IMultiTenantContainer _multitenantContainer;

    /// <summary>
    /// Constructs a <see cref="MultiTenantServiceProviderFactory"/>.
    /// </summary>
    /// <param name="multitenantContainer">The tenant distributor.</param>
    public MultiTenantServiceProviderFactory(MultiTenantContainer multitenantContainer)
    {
        _multitenantContainer = multitenantContainer;
    }

    /// <inheritdoc />
    public IMultiTenantContainer CreateBuilder(IServiceCollection services)
    {
        var container = services.CreateBuilder(_multitenantContainer.RootContainer);
        container.RegisterInstance(_multitenantContainer);
        container.ReMap<IServiceScopeFactory>(c => c.WithFactory(r => new StashboxServiceScopeFactory(r)));
        return _multitenantContainer;
    }

    /// <inheritdoc />
    public IServiceProvider CreateServiceProvider(IMultiTenantContainer containerBuilder)
    {
        return new StashboxServiceProvider(containerBuilder.RootContainer);
    }
}
```

**Comment by z4kn4fein** [comment]
Hey @vladimir-ilnytskyi, thank you for the snippets! I've released a new version of the main package (`v5.17.0`) and of the extensions (`v5.7.0`) that should fix this issue. `IEnumerable<>` services are resolved in the correct order now.

Also, based on your usage, you might find the [`Stashbox.AspNetCore.Multitenant`](https://github.com/z4kn4fein/stashbox-extensions-dependencyinjection#multitenant) package interesting. It is specifically designed for these kinds of multitenant scenarios. Let me know what you think!

**Comment by vladimir-ilnytskyi** [comment]
Thanks a lot for the fast fix! Really appriciate how fast you reply comparing to other public repo owners =)

I remember i have tried to adopt this package but there are certain things that didn't fit:

1) Our tenants are not known before first request, so basically that first call will load tenant from tenant service where we store some tenant configuration based on which there might be different configurations involved (basically we use same service set for all of them) so tenant container is needed to have service options per database basically, and some per-tenant redis caching involved as well
2) tenant configuration often changes, so i need a mechanism how to fetch new configuration + inform all IOptionsCache<T> to invalidate cached version and force recreate new instance.
<img width="690" alt="image" src="https://github.com/user-attachments/assets/91872d0b-5598-41f3-995f-62b4fb7846cf" />
